nss: add option to use p11-kit

rnhmjoj opened the pull request:
###### Motivation for this change

The original motivation is https://github.com/NixOS/nixpkgs/issues/8247

After multiple attempts it seems I have finally understood how to use p11-kit-trust.
So far I've only built nss and tested the system trust store is used by qtwebengine.
Firefox and other Mozilla programs should be tested as well.


This commit adds an option to replace libnssckbi with the p11-kit-trust[1] module. It makes all NSS application (like Firefox, Chromium, etc.) use the system trust store (/etc/ssl/certs/ in NixOS) and other PKCS#11 modules without ad-hoc configuration. Enabling the option will additionally generate a p11-kit compatible
module of the Mozilla CA bundle.

This approach was first implemented in Fedora[2] and other distributions like Arch Linux, later. The certdata2pem.py script is in fact taken, and some of the build code adapted, from the ca-certificates Fedora package[3].

[1]: https://p11-glue.github.io/p11-glue/p11-kit/manual/trust-nss.html

[2]: https://fedoraproject.org/wiki/Features/SharedSystemCertificates

[3]: https://src.fedoraproject.org/rpms/ca-certificates/tree/master

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS
   - [x] firefox
   - [x] chromium
- [ ] Tested compilation of all pkgs that depend on this change (mass rebuild!)
- [x] Tested execution of all binary files (none)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

rnhmjoj committed:
nss: add option to use p11-kit

This commit adds an option to replace libnssckbi with the
p11-kit-trust[1] module. It makes all NSS application (like Firefox,
Chromium, etc.) use the system trust store (/etc/ssl/certs/ in NixOS)
and other PKCS#11 modules without ad-hoc configuration.

This approach was first implemented in Fedora[2] and other distributions
like Arch Linux, later.
[1]: https://p11-glue.github.io/p11-glue/p11-kit/manual/trust-nss.html

[2]: https://fedoraproject.org/wiki/Features/SharedSystemCertificates

ajs124 commented:
I might be mixing things up, but [certdata2pem](https://github.com/NixOS/nixpkgs/blob/master/pkgs/data/misc/cacert/default.nix#L10) is already how nixpkgs/nixos is getting its ca certs. From nss. So why extract them here again?

rnhmjoj commented:
Because I didn't know that: I thought cacert was a different CA bundle. I'll remove that part, thank you.

rnhmjoj commented:
Done. I removed the (duplicated) part.

rnhmjoj commented:
Since there are no maintainers I'm pinging by the history. @ajs124, can you take a look?

rnhmjoj commented:
ping

rnhmjoj commented:
So, do you have any objection? 

I'd like for nss to use the system trust store in `/etc/ssl/certs`. That way is possibile to take control of the validation in chromium, firefox and basically all other browsers by setting `security.pki` in NixOS. I can't think of any unwanted side effects of this change.

ajs124 commented:
No objections, but I also don't feel comfortable approving/merging this, because I don't feel like I know enough about nss and/or p11-kit.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-in-distress/3604/35

andir commented on pkgs/development/libraries/nss/default.nix:
```diff
@@ -126,6 +133,11 @@ in stdenv.mkDerivation rec {
 
+  postInstall = stdenv.lib.optionalString useP11kit ''
+    # Replace built-in trust with p11-kit connection
+    ln -sf ${p11-kit}/lib/pkcs11/p11-kit-trust.so $out/lib/libnssckbi.so
```
So, how do we gurantee that the ABI between this "fake" output and the one defined in the header files matches? Is that defined in the PKCS#11 standard?

rnhmjoj replied:
It's a good question: the library is [designed](https://p11-glue.github.io/p11-glue/p11-kit/manual/trust-nss.html) to be used like this, so I assume it's ABI compatible with nss, but I'm not sure.

Even if it is, this could be an issue if nss introduces an ABI incompatible change and Nixpkgs is updated before p11-kit. It should be noted p11-kit is maintained and used by red hat, so it should be update quickly enough.

andir commented:
> This commit adds an option to replace libnssckbi with the p11-kit-trust[1](https://p11-glue.github.io/p11-glue/p11-kit/manual/trust-nss.html) module. It makes all NSS application (like Firefox, Chromium, etc.) use the system trust store (/etc/ssl/certs/ in NixOS) and other PKCS#11 modules without ad-hoc configuration. Enabling the option will additionally generate a p11-kit compatible
> module of the Mozilla CA bundle.

What was the situation before that? Did all the applications linking to NSS use the bundled trust store that came with NSS? I still wonder if there wouldn't be a better way to deal with this but I fear NSS doesn't really care much about system wide certs.

rnhmjoj commented:
> Did all the applications linking to NSS use the bundled trust store that came with NSS?

I think so, certainly they aren't using the system wide store.

> I still wonder if there wouldn't be a better way to deal with this but I fear NSS doesn't really care much about system wide certs.

Probably not, red hat/fedora was in the same situation and p11-kit is what they came up with.

SuperSandro2000 commented:
Can you please target staging with the mass rebuild?

SuperSandro2000 commented on pkgs/development/libraries/nss/default.nix:
```diff
@@ -1,4 +1,11 @@
+{ stdenv, fetchurl, nspr, perl, zlib
+, sqlite, ninja
+, darwin, fixDarwinDylibNames, buildPackages
+, p11-kit ? null
+, useP11kit ? true
```
```suggestion
, useP11kit ? true, p11-kit ? null
```

rnhmjoj replied:
Done.

rnhmjoj commented:
Updated and changed base to staging. Sorry for the noise.

SuperSandro2000 commented:
@ofborg eval

rnhmjoj committed:
nixos/tests/custom-ca: init

This is a NixOS test for the security.pki options.

rnhmjoj commented:
I added a test to check that the `security.pki` options are working in every major browser engine.
Programs based on openssl and gnutls should already be working, nss-based browsers needs this PR.

Building this on staging would probably take days, but you can avoid the rebuilds by running it from master and using `replaceRuntimeDependencies` as follows:
```
{ system.replaceRuntimeDependencies =
    let 
      nss-p11kit = pkgs.nss.overrideDerivation (old: {
        postInstall = ''
          # Replace built-in trust with p11-kit connection
          ln -sf ${pkgs.p11-kit}/lib/pkcs11/p11-kit-trust.so $out/lib/libnssckbi.so
        '';
      });
    in [{ original = pkgs.nss; replacement = nss-p11kit; }];
}
```

rnhmjoj commented:
@GrahamcOfBorg eval

SuperSandro2000 commented:
Staging is broken right now.

rnhmjoj commented:
I saw FRidh pushed a commit to fix the eval but it's still broken (differently).

FRidh commented:
yes another fix in e3961ff41f0f6f6cb0927bcbc7235727eed288bf

rnhmjoj commented:
@GrahamcOfBorg test custom-ca

SuperSandro2000 commented on pkgs/development/libraries/nss/default.nix:
```diff
@@ -1,9 +1,14 @@
+{ stdenv, fetchurl, nspr, perl, zlib
+, sqlite, ninja
+, darwin, fixDarwinDylibNames, buildPackages
+, useP11kit ? true, p11-kit ? null
```
```suggestion
, useP11kit ? true, p11-kit
```

rnhmjoj replied:
Is there an RFC or docs that recommends against asserts for optional dependencies?

FRidh replied:
there is no documented guideline, however, in some cases they have been removed because they broke overriding.
b6187aa5967e4ed33809d2916ef9398d4a9d6af9 #73102

rnhmjoj replied:
Thank you.

SuperSandro2000 commented on pkgs/development/libraries/nss/default.nix:
```diff
@@ -1,9 +1,14 @@
   enableFIPS ? false
 }:
 
+assert useP11kit -> p11-kit != null;
+
```
```suggestion
```

rnhmjoj commented:
Thanks to the test, I'm now pretty confident this PR won't wreak havoc on Nixpkgs.
If you have no more changes to suggest I'd go on and try to merge it.

tfc commented:
Wow, the integration test is awesome. Thanks for that!

rnhmjoj commented:
> Wow, the integration test is awesome. Thanks for that!

Thank you!

I have fixed the conflict (stdenv.lib -> lib change) and rebased. I tested again `nixosTests.firefox`, `nixosTests.chromium` and `nixosTests.custom-ca` and everything still looks ok. I'll try merge it now.

flokli commented:
@rnhmjoj the commit message says:

```
This commit adds an option to replace libnssckbi with the
p11-kit-trust[1] module. It makes all NSS application (like Firefox,
Chromium, etc.) use the system trust store (/etc/ssl/certs/ in NixOS)
and other PKCS#11 modules without ad-hoc configuration.
```

It however seems p11-kit doesn't properly discover PKCS#11 modules on NixOS. I opened https://github.com/NixOS/nixpkgs/issues/171978.
